scripts/assign_taxonomy_from_blast/ClassifyContigNR.py

```diff
@@ -126,193 +152,271 @@ def map_gids_binary(gids, mapping_file):
+        dWeight = sum(collate.values())
+        sortCollate = sorted(collate.items(), key=operator.itemgetter(1), reverse=True)
+        nL = len(collate)  ## Number of hits collated at depth
+        if nL > 0:
+            dP = 0.0
+            ## dWeight is the sum of weights at this depth
+            if dWeight > 0.0:
+                ## dP is the weight of the best hit at depth, divided by the total weight at depth
+                dP = float(sortCollate[0][1]) / dWeight
+                ## If the dP for the best hit is larger than min_fraction then assign at this depth and map back
+                if dP > min_fraction:
+                    geneAssign[gene][depth] = (sortCollate[0][0], dP)
+                    ## Get back lineage for the best assignment
+                    assignBack = mapBack[depth][sortCollate[0][0]]
+                    depth2 = depth - 1
+                    for assignB in assignBack:
+                        geneAssign[gene][depth2] = (assignB, 1.0)
+                        depth2 -= 1
+                    break
+
                 else:
-                    geneAssign[gene][depth] = ('No hits', -1.0)
+                    geneAssign[gene][depth] = ('Unclassified', -1.0)
+            ## If dWeight is 0.0, set 'Unclassified'
             else:
-                geneAssign[gene][depth] = ('No hits',-1.0)
+                geneAssign[gene][depth] = ('Unclassified', -1.0)
+        ## If no hits collated at depth, set 'Unclassified'
+        else:
+            geneAssign[gene][depth] = ('Unclassified', -1.0)
+    return geneAssign
 
-    
-    with open(args.output_dir+"_genes.csv", "w") as text_file:
-        for gene in geneAssign.keys():
-            text_file.write('%s'%gene)
-            for depth in range(7):
-                (assign,p) = geneAssign[gene][depth]
-                text_file.write(',%s->%.3f'%(assign,p))
-            text_file.write('\n')
-            text_file.flush()
-
-    #collate contig predictions 
-    #import ipdb; ipdb.set_trace()
-    
+
+def assign_unclassified():
+    d = {}
+    for depth in range(7): d[depth] = ('Unclassified',-1.0)
+    return d
+
+def assign_taxonomy(matches, mapping, mapBack, lineages, lengths, contigGenes, min_fraction, taxa_identities):
+    geneAssign = defaultdict(dict)
     contigAssign = defaultdict(dict)
-    
     for contig, genes in contigGenes.items():
-    
         collate_hits = list()
         for depth in range(7):
             collate_hits.append(Counter())
-    
         for gene in genes:
-        
+            ## Perform taxonomic assignments for all genes on contig
+            try:
+                matchs = matches[gene]
+                collated_gene_hits = collate_gene_hits(matchs, mapping, lineages, taxa_identities)
+                geneAssign = make_assignment(geneAssign, gene, collated_gene_hits, mapBack, min_fraction)
+            except KeyError:
+                logging.info("No matches found for "+gene)
```

Comment:
Could this be a bit overwhelming if the input data is large? Perhaps hide behind a --verbose option or similiar?